### PR TITLE
refactor: refactor crawler

### PR DESCRIPTION
**Pull request by bolinocroustibat**
Closes https://github.com/datagouv/hydra/issues/146.

Code in `crawl.py` is bloated and thus difficult to read, understand and maintain.
This PR refactors the crawler methods to be more readable, easier to understand and more maintainable, by:

- Adding and improving docstrings for crawl-related methods:
  - Split file `crawl.py` into several files in a dedicated `crawl` directory
  - Split method `check_urls` into methods dedicated to selecting resources in a `select_batch_resources_to_check.py` file with its submethods, and  the `check_resources` method
  - Split method `compute_check_has_changed` into 3 methods: `process_check_data`, `has_check_changed` and `send_check_to_udata`
  - Adding a CRUD method `Check.get_by_resource_id` and use that one in crawler instead of direct SQL in crawler
  - Renaming the CRUD method `Check.get` into `Check.get_by_id` to avoid any confusion
  - Add a `with_deleted` argument to `Check.get_...` methods for more code clarity and flexibility
  - Move the method `get_excluded_clause` as a static method of `db.Resource` class
  - Change `process_resource` and `Resource.STATUSES` function naming and statuses namings
- Add and improve methods docstrings
- Add type hints
- Fix some methods type hints and return cases

**Question**: should we go further and rename the `/crawl` directory to `/checking` to comply with the new namings?

### CODE REVIEW

**Review by maudetes — COMMENTED** [review]
I think I like the naming of `batch` that we may use in naming. Some of the functions naming are imperfect but can be discussed live maybe for ideas?

File: udata_hydra/crawl/__init__.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+                    worker_priority="low",
+                )
+            )
+        for task in asyncio.as_completed(tasks):
+            result = await task
+            results[result] += 1
+            context.monitor().refresh(results)
+
+
+async def check_catalog(iterations: int = -1) -> None:
```
[the remark this answers is not in the document]
Comment (bolinocroustibat) [reply udata_hydra/crawl/__init__.py:43]: Agreed. Changed to start_checks(). Any suggestion welcome

File: udata_hydra/crawl/__init__.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+from udata_hydra.logger import setup_logging
+from udata_hydra.utils import queue  # noqa
+
+results: defaultdict = defaultdict(int)
+
+
+log = setup_logging()
+
+
+async def check_resources(to_parse: list[str]) -> None:
```
[the remark this answers is not in the document]
Comment (bolinocroustibat) [reply udata_hydra/crawl/__init__.py:21]: That makes sense. The idea was to have all things related to starters (everything related to several resources) in __init__.py, and only a the single-resource check logic in `check_resource.py`.
But your remark, I agree, I just changed the organization and naming. LMKWYT

File: udata_hydra/db/check.py [hunk 3]
```diff
@@ -12,20 +14,37 @@ class Check:
...
         pool = await context.pool()
         async with pool.acquire() as connection:
             q = """
                 SELECT * FROM catalog JOIN checks
                 ON catalog.last_check = checks.id
-                WHERE checks.id = $1;
+                WHERE checks.id = $1
             """
+            if not with_deleted:
+                q += " AND catalog.deleted = FALSE"
```
[the remark this answers is not in the document]
Comment (bolinocroustibat) [reply udata_hydra/db/check.py:26]: Until now, tests were using all checks including the deleted ones.
In the future I'd like to use that for testing the deletion, and for the other tests not to select the deleted ones anymore.

**Review by maudetes — APPROVED** [review]
I think we're good to go with this refactoring! Thank you :pray:

**Pull request merged**